consensus/cryptarchia-engine/src/lib.rs

```diff
@@ -252,17 +287,100 @@ where
+        let mut removed_blocks = vec![];
+        for prunable_fork_info in prunable_forks {
+            removed_blocks.extend(self.prune_fork(&prunable_fork_info));
+        }
+        removed_blocks.into_iter()
+    }
+
+    /// Get an iterator over the forks that can be pruned given the provided
+    /// depth.
+    ///
+    /// This means that all forks that diverged from the canonical chain before
+    /// the provided `depth` height are returned.
+    pub fn prunable_forks(&self, depth: u64) -> impl Iterator<Item = ForkDivergenceInfo<Id>> {
+        let local_chain = self.local_chain;
+        let Some(target_height) = local_chain.length.checked_sub(depth) else {
+            tracing::info!(
+                target: LOG_TARGET,
+                "No prunable fork, the canonical chain is not longer than the provided depth. Canonical chain length: {}, provided depth: {}", local_chain.length, depth
+            );
+            return vec![].into_iter();
+        };
+        self.non_canonical_forks()
+            .filter_map(|fork| {
+                // We calculate LCA once and store it in `ForkInfo` so it can be consumed
+                // elsewhere without the need to re-calculate it.
+                let lca = self.branches.lca(&local_chain, &fork);
+                (lca.length <= target_height).then_some(ForkDivergenceInfo { tip: fork, lca })
+            })
+            .collect::<Vec<_>>()
+            .into_iter()
+    }
+
+    /// Returns all the forks that are not part of the local canonical chain.
+    ///
+    /// The result contains both prunable and non prunable forks.
+    pub fn non_canonical_forks(&self) -> impl Iterator<Item = Branch<Id>> + '_ {
+        self.branches
+            .branches()
+            .filter(|fork_tip| fork_tip.id != self.tip())
+    }
+
+    /// Remove all blocks from `tip` to `lca`, excluding `lca` which belongs to
+    /// the canonical chain.
+    fn prune_fork(
+        &mut self,
+        &ForkDivergenceInfo { lca, tip }: &ForkDivergenceInfo<Id>,
+    ) -> impl Iterator<Item = Id> {
+        let tip_removed = self.branches.tips.remove(&tip.id);
+        if !tip_removed {
+            tracing::error!(target: LOG_TARGET, "Fork tip {tip:#?} not found in the set of tips.");
+        }
+
+        let mut current_tip = tip.id;
+        let mut removed_blocks = vec![];
+        while current_tip != lca.id {
+            let Some(branch) = self.branches.branches.remove(&current_tip) else {
+                // If tip is not in branch set, it means this tip was sharing part of its
+                // history with another fork that has already been removed.
+                break;
+            };
+            removed_blocks.push(branch.id);
+            current_tip = branch.parent;
+        }
+        tracing::debug!(
+            target: LOG_TARGET,
+            "Pruned {} branches from {tip:#?} to {current_tip:#?}.", removed_blocks.len()
+        );
+        removed_blocks.into_iter()
```

Comment:
Using a vec here is ok probably. But I'm pretty sure we could avoid it (this looks like a reduce....like everything 😂 )

Reply:
Mhh we have a `while` condition, and we are removing nodes from the storage map, I could not think of an alternative way of doing this, but I am of course open to suggestions!

Reply:
> Mhh we have a `while` condition, and we are removing nodes from the storage map, I could not think of an alternative way of doing this, but I am of course open to suggestions!

Nah, it would be too much work to fit everything and its more readable this way. Hopefully in some not so far future we get the generators syntax (which could help on this kind of things I think).